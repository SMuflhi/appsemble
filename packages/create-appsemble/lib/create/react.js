import path from 'path';

import fs from 'fs-extra';
import sortPackageJson from 'sort-package-json';

export default async function react(outputPath, pkg) {
  const { dependencies } = await fs.readJSON(
    path.resolve(__dirname, '../../templates/react/package.json'),
  );

  const outPkg = sortPackageJson({
    ...pkg,
    dependencies: { ...pkg.dependencies, ...dependencies },
  });

  await fs.copy(path.resolve(__dirname, '../../templates/react'), outputPath);
  await fs.outputJSON(path.join(outputPath, 'package.json'), outPkg, { spaces: 2 });
}
