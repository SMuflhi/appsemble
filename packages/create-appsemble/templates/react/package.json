{
  "dependencies": {
    "@appsemble/react-bulma": "^1.0.0",
    "@appsemble/react": "^1.0.0",
    "@appsemble/sdk": "^1.0.0",
    "react": "^16.7.0"
  }
}
