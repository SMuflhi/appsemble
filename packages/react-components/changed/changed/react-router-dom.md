Upgrade from “react-router-dom” 5.0.0 to 6.0.0
