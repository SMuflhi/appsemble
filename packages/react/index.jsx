import { bootstrap as sdkBootstrap } from '@appsemble/sdk';
import React from 'react';
import { render } from 'react-dom';
import retargetEvents from 'react-shadow-dom-retarget-events';

const { Consumer, Provider } = React.createContext();

/**
 * Mount a React component returned by a bootstrap function in the shadow DOM of a block.
 */
export function mount(Component, root) {
  return params => {
    const reactRoot = params.shadowRoot.appendChild(
      root ? root.cloneNode() : document.createElement('div'),
    );
    const props = {
      ...params,
      reactRoot,
    };
    const component = (
      <Provider value={props}>
        <Component {...props} />
      </Provider>
    );
    render(component, reactRoot);
    /**
     * React doesnâ€™t play nice with shadow DOM. This library works around that. However, the
     * implementation does contain some bugs.
     *
     * Related issues:
     *
     * - https://github.com/facebook/react/issues/2043
     * - https://github.com/facebook/react/issues/9242
     * - https://github.com/facebook/react/issues/10422
     * - https://github.com/facebook/react/issues/12973
     * - https://github.com/facebook/react/pull/8117
     * - https://github.com/facebook/react/pull/12163
     *
     * @todo Remove this when React has fixes this issue.
     */
    retargetEvents(params.shadowRoot);
  };
}

export function bootstrap(Component, reactRoot) {
  return sdkBootstrap(mount(Component, reactRoot));
}

/**
 * A HOC which passes the Appsemble block values to he wrapped React component.
 */
export function withBlock(Component) {
  function Wrapper(props) {
    return <Consumer>{values => <Component {...values} {...props} />}</Consumer>;
  }
  if (process.env.NODE_ENV !== 'production') {
    Wrapper.displayName = `withBlock(${Component.displayName || Component.name})`;
  }
  return Wrapper;
}
