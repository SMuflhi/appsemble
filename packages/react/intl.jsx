import React from 'react';
import {
  IntlProvider,
} from 'react-intl';


/**
 * A HOC which provides `intl` in a block written in React.
 */
// eslint-disable-next-line import/prefer-default-export
export function provideIntl(Component) {
  return class extends React.Component {
    render() {
      return (
        <IntlProvider locale="en-US" defaultLocale="en-US" textComponent={React.Fragment}>
          <Component {...this.props} />
        </IntlProvider>
      );
    }
  };
}
