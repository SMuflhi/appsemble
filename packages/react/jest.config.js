module.exports = {
  displayName: '@appsemble/react',
  moduleNameMapper: {
    [/@appsemble\/([\w-]+)/.source]: '@appsemble/$1/src',
  },
  setupFilesAfterEnv: ['<rootDir>/jest.setup.ts'],
  snapshotSerializers: ['enzyme-to-json/serializer'],
};
