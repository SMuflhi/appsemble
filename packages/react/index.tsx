import { bootstrap as sdkBootstrap, BootstrapParams } from '@appsemble/sdk';
import * as React from 'react';
import * as ReactDOM from 'react-dom';
import retargetEvents from 'react-shadow-dom-retarget-events';

export interface BlockProps extends BootstrapParams {
  /**
   * The DOM node on which the block is mounted.
   */
  reactRoot: HTMLElement;
}

const { Consumer, Provider } = React.createContext<BlockProps>(null);

/**
 * Mount a React component returned by a bootstrap function in the shadow DOM of a block.
 */
export function mount(
  Component: React.ComponentType<BlockProps>,
  root?: HTMLElement,
): (params: BootstrapParams) => void {
  return params => {
    const reactRoot = params.shadowRoot.appendChild(
      root ? root.cloneNode() : document.createElement('div'),
    ) as HTMLElement;
    const props = {
      ...params,
      reactRoot,
    };
    const component: JSX.Element = (
      <Provider value={props}>
        <Component {...props} />
      </Provider>
    );
    ReactDOM.render(component, reactRoot);
    params.utils.addCleanup(() => ReactDOM.unmountComponentAtNode(reactRoot));
    /**
     * React doesnâ€™t play nice with shadow DOM. This library works around that. However, the
     * implementation does contain some bugs.
     *
     * Related issues:
     *
     * - https://github.com/facebook/react/issues/2043
     * - https://github.com/facebook/react/issues/9242
     * - https://github.com/facebook/react/issues/10422
     * - https://github.com/facebook/react/issues/12973
     * - https://github.com/facebook/react/pull/8117
     * - https://github.com/facebook/react/pull/12163
     *
     * @todo Remove this when React has fixes this issue.
     */
    retargetEvents(params.shadowRoot);
  };
}

export function bootstrap(
  Component: React.ComponentType<BlockProps>,
  reactRoot?: HTMLElement,
): void {
  sdkBootstrap(mount(Component, reactRoot));
}

/**
 * A HOC which passes the Appsemble block values to he wrapped React component.
 */
export function withBlock<P extends object>(
  Component: React.ComponentType<P & Omit<BlockProps, keyof P>>,
): React.ComponentType<P> {
  function Wrapper(props: P): JSX.Element {
    return <Consumer>{values => <Component {...values} {...props} />}</Consumer>;
  }
  if (process.env.NODE_ENV !== 'production') {
    Wrapper.displayName = `withBlock(${Component.displayName || Component.name})`;
  }
  return Wrapper;
}
