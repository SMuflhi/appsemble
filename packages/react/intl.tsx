import * as React from 'react';
import { IntlProvider } from 'react-intl';

/**
 * A HOC which provides `intl` in a block written in React.
 */
// eslint-disable-next-line import/prefer-default-export
export function provideIntl(
  Component: React.ComponentType,
  intlProviderProps: IntlProvider.Props,
): React.ComponentType {
  return props => (
    <IntlProvider
      defaultLocale="en-US"
      locale="en-US"
      {...intlProviderProps}
      textComponent={React.Fragment}
    >
      <Component {...props} />
    </IntlProvider>
  );
}
