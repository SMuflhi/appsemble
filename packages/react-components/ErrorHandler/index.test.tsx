import { render } from '@testing-library/react';
import { type ReactElement } from 'react';
import { BrowserRouter } from 'react-router-dom';

import { ErrorHandler } from './index.js';

it('should render its children if no errors are thrown', () => {
  function Child(): ReactElement {
    return <p>Test</p>;
  }
  function Fallback(): ReactElement {
    return <p>Something went wrong!</p>;
  }
  const { container } = render(
    <BrowserRouter>
      {/* eslint-disable-next-line react/jsx-no-bind */}
      <ErrorHandler fallback={Fallback}>
        <Child />
      </ErrorHandler>
    </BrowserRouter>,
  );

  expect(container).toMatchSnapshot();
});

it('should render its fallback when errors are thrown', () => {
  vi.spyOn(console, 'error').mockImplementation(null);
  const Child = (): ReactElement => {
    throw new Error('test');
  };
  function Fallback(): ReactElement {
    return <p>Something went wrong!</p>;
  }
  const { container } = render(
    <BrowserRouter>
      {/* eslint-disable-next-line react/jsx-no-bind */}
      <ErrorHandler fallback={Fallback}>
        <Child />
      </ErrorHandler>
    </BrowserRouter>,
  );

  expect(container).toMatchSnapshot();
});
