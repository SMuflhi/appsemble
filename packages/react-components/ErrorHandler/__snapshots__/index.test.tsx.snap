// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should render its children if no errors are thrown 1`] = `
<div>
  <p>
    Test
  </p>
</div>
`;

exports[`should render its fallback when errors are thrown 1`] = `
<div>
  <p>
    Something went wrong!
  </p>
</div>
`;
